File: go/streaming/video_service.go
// SPDX-License-Identifier: MIT

package streaming

import (
	"bytes"
	"context"
	"encoding/binary"
	"encoding/json"
	"errors"
	"fmt"
	"sync"
	"time"

	"github.com/google/uuid"
	"github.com/thegeeknetwork/aether-protocol-go/constants"
	"github.com/thegeeknetwork/aether-protocol-go/protocol"
	"github.com/thegeeknetwork/aether-protocol-go/routing"
)

// ──────────────────────────────────────────────
// JSON signaling types
// ──────────────────────────────────────────────

// VideoSignalingMessage is the JSON-encoded payload for VideoSignaling packets.
type VideoSignalingMessage struct {
	Kind           string   `json:"kind"`
	CallID         string   `json:"call_id"`
	FromUhid       string   `json:"from_uhid"`
	ToUhid         string   `json:"to_uhid"`
	ProposedCodecs []string `json:"proposed_codecs,omitempty"`
	SelectedCodec  string   `json:"selected_codec,omitempty"`
	Width          int      `json:"width,omitempty"`
	Height         int      `json:"height,omitempty"`
	FPS            int      `json:"fps,omitempty"`
	BitrateKbps    int      `json:"bitrate_kbps,omitempty"`
	Reason         string   `json:"reason,omitempty"`
}

// ──────────────────────────────────────────────
// Session state
// ──────────────────────────────────────────────

// VideoCallState mirrors the voice call state machine for video.
type VideoCallState int

const (
	VideoCallStateOffering VideoCallState = iota
	VideoCallStateRinging
	VideoCallStateActive
	VideoCallStateEnded
)

// VideoCallSession holds per-call runtime state.
type VideoCallSession struct {
	CallID    uuid.UUID
	PeerUhid  string
	State     VideoCallState
	Sequence  uint32
	CreatedAt time.Time
}

// ──────────────────────────────────────────────
// Service
// ──────────────────────────────────────────────

// VideoCallService manages unicast video calls over the mesh.
type VideoCallService struct {
	sender    routing.MeshSender
	localUhid string
	calls     sync.Map // uuid.UUID -> *VideoCallSession

	// Event callbacks — optional.
	OnCallOffered      func(session *VideoCallSession)
	OnCallAccepted     func(session *VideoCallSession)
	OnCallEnded        func(session *VideoCallSession, reason string)
	OnFrameReceived    func(session *VideoCallSession, frame *VideoFrameData)
	OnKeyframeRequired func(session *VideoCallSession)
	OnQualityChanged   func(session *VideoCallSession, msg *VideoSignalingMessage)
}

// VideoFrameData is the decoded form of a binary VideoFrame payload.
type VideoFrameData struct {
	CallID         uuid.UUID
	Sequence       uint32
	TimestampMs    int64
	IsKeyframe     bool
	EncodedPayload []byte
}

// NewVideoCallService constructs a VideoCallService.
func NewVideoCallService(sender routing.MeshSender) *VideoCallService {
	if sender == nil {
		panic("streaming: video sender must not be nil")
	}
	return &VideoCallService{
		sender:    sender,
		localUhid: sender.LocalUhid(),
	}
}

// SendOffer initiates a video call to toUhid. Returns the new call ID.
func (s *VideoCallService) SendOffer(ctx context.Context, toUhid string, codecs []string, width, height, fps, bitrateKbps int) (uuid.UUID, error) {
	if toUhid == "" {
		return uuid.Nil, errors.New("streaming: toUhid must not be empty")
	}
	callID := uuid.New()
	session := &VideoCallSession{
		CallID:    callID,
		PeerUhid:  toUhid,
		State:     VideoCallStateOffering,
		CreatedAt: time.Now(),
	}
	s.calls.Store(callID, session)

	msg := VideoSignalingMessage{
		Kind:           "offer",
		CallID:         callID.String(),
		FromUhid:       s.localUhid,
		ToUhid:         toUhid,
		ProposedCodecs: codecs,
		Width:          width,
		Height:         height,
		FPS:            fps,
		BitrateKbps:    bitrateKbps,
	}
	if err := s.sendSignaling(ctx, toUhid, msg); err != nil {
		s.calls.Delete(callID)
		return uuid.Nil, err
	}
	return callID, nil
}

// AcceptCall sends an "answer" signaling message for an inbound video call.
func (s *VideoCallService) AcceptCall(ctx context.Context, callID uuid.UUID) error {
	session, err := s.getVideoSession(callID)
	if err != nil {
		return err
	}
	if session.State != VideoCallStateRinging {
		return fmt.Errorf("streaming: video call %s is not in ringing state", callID)
	}
	session.State = VideoCallStateActive

	msg := VideoSignalingMessage{
		Kind:     "answer",
		CallID:   callID.String(),
		FromUhid: s.localUhid,
		ToUhid:   session.PeerUhid,
	}
	return s.sendSignaling(ctx, session.PeerUhid, msg)
}

// HangUp ends a video call and notifies the peer.
func (s *VideoCallService) HangUp(ctx context.Context, callID uuid.UUID) error {
	session, err := s.getVideoSession(callID)
	if err != nil {
		return err
	}
	session.State = VideoCallStateEnded

	msg := VideoSignalingMessage{
		Kind:     "hangup",
		CallID:   callID.String(),
		FromUhid: s.localUhid,
		ToUhid:   session.PeerUhid,
	}
	if cb := s.OnCallEnded; cb != nil {
		cb(session, "local_hangup")
	}
	s.calls.Delete(callID)
	return s.sendSignaling(ctx, session.PeerUhid, msg)
}

// SendFrame encodes and transmits a single video frame for an active call.
//
// Binary layout (VideoFrame):
//
//	[16] CallId       (UUID, big-endian)
//	[4]  Sequence     (uint32, little-endian)
//	[8]  TimestampMs  (int64, little-endian)
//	[1]  IsKeyframe   (0 or 1)
//	[N]  EncodedPayload
func (s *VideoCallService) SendFrame(ctx context.Context, callID uuid.UUID, encodedVideo []byte, isKeyframe bool) error {
	session, err := s.getVideoSession(callID)
	if err != nil {
		return err
	}
	if session.State != VideoCallStateActive {
		return fmt.Errorf("streaming: video call %s is not active", callID)
	}

	session.Sequence++
	payload, err := marshalVideoFrame(callID, session.Sequence, time.Now().UnixMilli(), isKeyframe, encodedVideo)
	if err != nil {
		return err
	}

	pkt := protocol.NewMeshPacket()
	pkt.Type = protocol.VideoFrame
	pkt.SourceUhid = s.localUhid
	pkt.DestinationUhid = session.PeerUhid
	pkt.Ttl = constants.DefaultTtl
	pkt.Priority = 90 // video slightly below voice SOS
	pkt.Payload = payload

	_, err = s.sender.Send(ctx, pkt, session.PeerUhid)
	return err
}

// RequestKeyframe sends a "keyframe_request" signaling message to the peer.
func (s *VideoCallService) RequestKeyframe(ctx context.Context, callID uuid.UUID) error {
	session, err := s.getVideoSession(callID)
	if err != nil {
		return err
	}
	msg := VideoSignalingMessage{
		Kind:     "keyframe_request",
		CallID:   callID.String(),
		FromUhid: s.localUhid,
		ToUhid:   session.PeerUhid,
	}
	return s.sendSignaling(ctx, session.PeerUhid, msg)
}

// NotifyQualityChange sends a "quality_change" signaling message to inform the
// peer of the desired encoding parameters.
func (s *VideoCallService) NotifyQualityChange(ctx context.Context, callID uuid.UUID, width, height, fps, bitrateKbps int) error {
	session, err := s.getVideoSession(callID)
	if err != nil {
		return err
	}
	msg := VideoSignalingMessage{
		Kind:        "quality_change",
		CallID:      callID.String(),
		FromUhid:    s.localUhid,
		ToUhid:      session.PeerUhid,
		Width:       width,
		Height:      height,
		FPS:         fps,
		BitrateKbps: bitrateKbps,
	}
	return s.sendSignaling(ctx, session.PeerUhid, msg)
}

// HandlePacket processes inbound VideoSignaling or VideoFrame packets.
func (s *VideoCallService) HandlePacket(ctx context.Context, packet *protocol.MeshPacket) error {
	if packet == nil {
		return errors.New("streaming: video packet must not be nil")
	}
	switch packet.Type {
	case protocol.VideoSignaling:
		return s.handleVideoSignaling(ctx, packet)
	case protocol.VideoFrame:
		return s.handleVideoFrame(packet)
	default:
		return fmt.Errorf("streaming: unexpected video packet type %s", packet.Type)
	}
}

// ──────────────────────────────────────────────
// Internal helpers
// ──────────────────────────────────────────────

func (s *VideoCallService) handleVideoSignaling(_ context.Context, packet *protocol.MeshPacket) error {
	var msg VideoSignalingMessage
	if err := json.Unmarshal(packet.Payload, &msg); err != nil {
		return fmt.Errorf("streaming: unmarshal video signaling: %w", err)
	}
	callID, err := uuid.Parse(msg.CallID)
	if err != nil {
		return fmt.Errorf("streaming: invalid call_id %q: %w", msg.CallID, err)
	}

	switch msg.Kind {
	case "offer":
		session := &VideoCallSession{
			CallID:    callID,
			PeerUhid:  packet.SourceUhid,
			State:     VideoCallStateRinging,
			CreatedAt: time.Now(),
		}
		s.calls.Store(callID, session)
		if cb := s.OnCallOffered; cb != nil {
			cb(session)
		}

	case "answer":
		if session, err := s.getVideoSession(callID); err == nil {
			session.State = VideoCallStateActive
			if cb := s.OnCallAccepted; cb != nil {
				cb(session)
			}
		}

	case "hangup", "cancel", "timeout":
		if session, err := s.getVideoSession(callID); err == nil {
			session.State = VideoCallStateEnded
			if cb := s.OnCallEnded; cb != nil {
				cb(session, msg.Kind)
			}
			s.calls.Delete(callID)
		}

	case "keyframe_request":
		if session, err := s.getVideoSession(callID); err == nil {
			if cb := s.OnKeyframeRequired; cb != nil {
				cb(session)
			}
		}

	case "quality_change":
		if session, err := s.getVideoSession(callID); err == nil {
			if cb := s.OnQualityChanged; cb != nil {
				cb(session, &msg)
			}
		}
	}
	return nil
}

func (s *VideoCallService) handleVideoFrame(packet *protocol.MeshPacket) error {
	frame, err := unmarshalVideoFrame(packet.Payload)
	if err != nil {
		return fmt.Errorf("streaming: unmarshal video frame: %w", err)
	}
	session, err := s.getVideoSession(frame.CallID)
	if err != nil {
		return nil // unknown call — drop
	}
	if cb := s.OnFrameReceived; cb != nil {
		cb(session, frame)
	}
	return nil
}

func (s *VideoCallService) getVideoSession(callID uuid.UUID) (*VideoCallSession, error) {
	v, ok := s.calls.Load(callID)
	if !ok {
		return nil, fmt.Errorf("streaming: unknown video call %s", callID)
	}
	return v.(*VideoCallSession), nil
}

func (s *VideoCallService) sendSignaling(ctx context.Context, toUhid string, msg VideoSignalingMessage) error {
	body, err := json.Marshal(msg)
	if err != nil {
		return fmt.Errorf("streaming: marshal video signaling: %w", err)
	}
	pkt := protocol.NewMeshPacket()
	pkt.Type = protocol.VideoSignaling
	pkt.SourceUhid = s.localUhid
	pkt.DestinationUhid = toUhid
	pkt.Ttl = constants.DefaultTtl
	pkt.Payload = body
	_, err = s.sender.Send(ctx, pkt, toUhid)
	return err
}

// ──────────────────────────────────────────────
// Binary serialisation helpers
// ──────────────────────────────────────────────

// marshalVideoFrame encodes a VideoFrame into its canonical binary format.
func marshalVideoFrame(callID uuid.UUID, seq uint32, tsMs int64, isKeyframe bool, payload []byte) ([]byte, error) {
	var buf bytes.Buffer
	buf.Write(callID[:])
	if err := binary.Write(&buf, binary.LittleEndian, seq); err != nil {
		return nil, err
	}
	if err := binary.Write(&buf, binary.LittleEndian, tsMs); err != nil {
		return nil, err
	}
	if isKeyframe {
		buf.WriteByte(1)
	} else {
		buf.WriteByte(0)
	}
	buf.Write(payload)
	return buf.Bytes(), nil
}

// unmarshalVideoFrame decodes a binary VideoFrame payload.
func unmarshalVideoFrame(data []byte) (*VideoFrameData, error) {
	const fixedSize = 16 + 4 + 8 + 1
	if len(data) < fixedSize {
		return nil, fmt.Errorf("streaming: video frame too short: %d bytes", len(data))
	}
	r := bytes.NewReader(data)

	var callIDBytes [16]byte
	if _, err := r.Read(callIDBytes[:]); err != nil {
		return nil, err
	}
	callID, err := uuid.FromBytes(callIDBytes[:])
	if err != nil {
		return nil, err
	}

	var seq uint32
	if err := binary.Read(r, binary.LittleEndian, &seq); err != nil {
		return nil, err
	}

	var tsMs int64
	if err := binary.Read(r, binary.LittleEndian, &tsMs); err != nil {
		return nil, err
	}

	kfByte, err := r.ReadByte()
	if err != nil {
		return nil, err
	}

	encoded := make([]byte, r.Len())
	if _, err := r.Read(encoded); err != nil && r.Len() > 0 {
		return nil, err
	}

	return &VideoFrameData{
		CallID:         callID,
		Sequence:       seq,
		TimestampMs:    tsMs,
		IsKeyframe:     kfByte != 0,
		EncodedPayload: encoded,
	}, nil
}

File: rust/src/streaming/video_service.rs
// SPDX-License-Identifier: MIT

//! 1-to-1 video call service: signaling state machine + binary frame delivery.

use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use std::time::{SystemTime, UNIX_EPOCH};
use uuid::Uuid;

use crate::constants::DEFAULT_TTL;
use crate::protocol::{MeshPacket, PacketType};
use crate::routing::sender::MeshSender;
use crate::routing::RoutingService;
use tokio::sync::broadcast;

// ── Signaling wire type ────────────────────────────────────────────────────────

/// JSON signaling for video calls. Cross-language canonical wire format.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct VideoSignalingMessage {
    pub kind: String,
    pub call_id: Uuid,
    pub from_uhid: String,
    pub to_uhid: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub proposed_codecs: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub selected_codec: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub width: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub height: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub fps: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub bitrate_kbps: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reason: Option<String>,
}

// ── Call state machine ─────────────────────────────────────────────────────────

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum VideoCallState {
    Outgoing,
    Incoming,
    Connected,
    Ended,
    Failed,
}

#[derive(Debug, Clone)]
pub struct VideoCallEntry {
    pub call_id: Uuid,
    pub remote_uhid: String,
    pub state: VideoCallState,
}

// ── Service ───────────────────────────────────────────────────────────────────

/// Capacity of the inbound-frame broadcast channel; lagging subscribers drop oldest.
const FRAME_RECEIVED_CHANNEL_CAPACITY: usize = 256;

/// An inbound video/call frame decoded from a received VIDEO_FRAME/VIDEO_CALL packet,
/// delivered to subscribers of [`VideoCallService::subscribe_frames`].
#[derive(Debug, Clone)]
pub struct VideoFrameReceivedEvent {
    pub call_id: Uuid,
    pub sender_uhid: String,
    pub seq: u32,
    pub timestamp_ms: i64,
    pub is_keyframe: bool,
    pub data: Vec<u8>,
}

pub struct VideoCallService {
    sender: Arc<dyn MeshSender>,
    routing: Arc<RoutingService>,
    calls: Mutex<HashMap<Uuid, VideoCallEntry>>,
    frame_received_tx: broadcast::Sender<VideoFrameReceivedEvent>,
}

impl VideoCallService {
    pub fn new(sender: Arc<dyn MeshSender>, routing: Arc<RoutingService>) -> Self {
        let (frame_received_tx, _) = broadcast::channel(FRAME_RECEIVED_CHANNEL_CAPACITY);
        Self {
            sender,
            routing,
            calls: Mutex::new(HashMap::new()),
            frame_received_tx,
        }
    }

    /// Subscribe to inbound video/call frames decoded from received VIDEO_FRAME /
    /// VIDEO_CALL packets.
    pub fn subscribe_frames(&self) -> broadcast::Receiver<VideoFrameReceivedEvent> {
        self.frame_received_tx.subscribe()
    }

    /// Originate a video call to `to_uhid`. Returns the new call-id.
    pub async fn send_offer(
        &self,
        to_uhid: &str,
        codecs: Vec<String>,
        width: u32,
        height: u32,
        fps: u32,
        bitrate_kbps: u32,
    ) -> Result<Uuid, String> {
        if to_uhid.is_empty() {
            return Err("to_uhid is empty".into());
        }
        let call_id = Uuid::new_v4();
        {
            let mut calls = self.calls.lock().unwrap();
            calls.insert(
                call_id,
                VideoCallEntry {
                    call_id,
                    remote_uhid: to_uhid.to_string(),
                    state: VideoCallState::Outgoing,
                },
            );
        }
        let sig = VideoSignalingMessage {
            kind: "offer".into(),
            call_id,
            from_uhid: self.sender.local_uhid(),
            to_uhid: to_uhid.to_string(),
            proposed_codecs: Some(codecs),
            selected_codec: None,
            width: Some(width),
            height: Some(height),
            fps: Some(fps),
            bitrate_kbps: Some(bitrate_kbps),
            reason: None,
        };
        self.send_signaling(to_uhid, &sig).await;
        Ok(call_id)
    }

    /// Accept an incoming video call.
    pub async fn accept_call(&self, call_id: Uuid) -> Result<(), String> {
        let remote_uhid = {
            let mut calls = self.calls.lock().unwrap();
            let entry = calls
                .get_mut(&call_id)
                .ok_or_else(|| format!("unknown call_id {}", call_id))?;
            if entry.state != VideoCallState::Incoming {
                return Err(format!("call {} is not in Incoming state", call_id));
            }
            entry.state = VideoCallState::Connected;
            entry.remote_uhid.clone()
        };
        let sig = VideoSignalingMessage {
            kind: "answer".into(),
            call_id,
            from_uhid: self.sender.local_uhid(),
            to_uhid: remote_uhid.clone(),
            proposed_codecs: None,
            selected_codec: None,
            width: None,
            height: None,
            fps: None,
            bitrate_kbps: None,
            reason: None,
        };
        self.send_signaling(&remote_uhid, &sig).await;
        Ok(())
    }

    /// Hang up a video call.
    pub async fn hang_up(&self, call_id: Uuid) -> Result<(), String> {
        let remote_uhid = {
            let mut calls = self.calls.lock().unwrap();
            let entry = calls
                .get_mut(&call_id)
                .ok_or_else(|| format!("unknown call_id {}", call_id))?;
            let prev = entry.state.clone();
            entry.state = VideoCallState::Ended;
            if matches!(prev, VideoCallState::Ended | VideoCallState::Failed) {
                return Ok(());
            }
            entry.remote_uhid.clone()
        };
        let sig = VideoSignalingMessage {
            kind: "hangup".into(),
            call_id,
            from_uhid: self.sender.local_uhid(),
            to_uhid: remote_uhid.clone(),
            proposed_codecs: None,
            selected_codec: None,
            width: None,
            height: None,
            fps: None,
            bitrate_kbps: None,
            reason: None,
        };
        self.send_signaling(&remote_uhid, &sig).await;
        Ok(())
    }

    /// Send an encoded video frame.
    ///
    /// Binary payload wire format:
    ///   [16] CallId (UUID RFC4122 big-endian)
    ///   [4]  Sequence (u32 little-endian)
    ///   [8]  TimestampMs (i64 little-endian)
    ///   [1]  IsKeyframe (0 or 1)
    ///   [N]  EncodedPayload
    pub async fn send_frame(
        &self,
        call_id: Uuid,
        encoded_video: &[u8],
        is_keyframe: bool,
    ) -> Result<(), String> {
        let remote_uhid = {
            let calls = self.calls.lock().unwrap();
            let entry = calls
                .get(&call_id)
                .ok_or_else(|| format!("unknown call_id {}", call_id))?;
            if entry.state != VideoCallState::Connected {
                return Err(format!("call {} is not connected", call_id));
            }
            entry.remote_uhid.clone()
        };
        let payload = build_video_frame(call_id, encoded_video, is_keyframe);

        let mut pkt = MeshPacket::new(PacketType::VideoFrame, self.sender.local_uhid());
        pkt.destination_uhid = remote_uhid.clone();
        pkt.ttl = DEFAULT_TTL;
        pkt.priority = 64;
        pkt.payload = payload;
        self.sender.send(&pkt, &remote_uhid).await;
        Ok(())
    }

    /// Request a keyframe from the remote peer.
    pub async fn request_keyframe(&self, call_id: Uuid) -> Result<(), String> {
        let remote_uhid = {
            let calls = self.calls.lock().unwrap();
            calls
                .get(&call_id)
                .ok_or_else(|| format!("unknown call_id {}", call_id))?
                .remote_uhid
                .clone()
        };
        let sig = VideoSignalingMessage {
            kind: "keyframe_request".into(),
            call_id,
            from_uhid: self.sender.local_uhid(),
            to_uhid: remote_uhid.clone(),
            proposed_codecs: None,
            selected_codec: None,
            width: None,
            height: None,
            fps: None,
            bitrate_kbps: None,
            reason: None,
        };
        self.send_signaling(&remote_uhid, &sig).await;
        Ok(())
    }

    /// Notify the remote peer of a quality change.
    pub async fn notify_quality_change(
        &self,
        call_id: Uuid,
        width: u32,
        height: u32,
        fps: u32,
        bitrate_kbps: u32,
    ) -> Result<(), String> {
        let remote_uhid = {
            let calls = self.calls.lock().unwrap();
            calls
                .get(&call_id)
                .ok_or_else(|| format!("unknown call_id {}", call_id))?
                .remote_uhid
                .clone()
        };
        let sig = VideoSignalingMessage {
            kind: "quality_change".into(),
            call_id,
            from_uhid: self.sender.local_uhid(),
            to_uhid: remote_uhid.clone(),
            proposed_codecs: None,
            selected_codec: None,
            width: Some(width),
            height: Some(height),
            fps: Some(fps),
            bitrate_kbps: Some(bitrate_kbps),
            reason: None,
        };
        self.send_signaling(&remote_uhid, &sig).await;
        Ok(())
    }

    /// Route an inbound packet to the right handler.
    pub async fn handle_packet(&self, packet: &MeshPacket) -> Result<(), String> {
        match packet.packet_type {
            PacketType::VideoSignaling => self.handle_signaling(packet).await,
            PacketType::VideoFrame | PacketType::VideoCall => self.handle_video_frame(packet).await,
            _ => Ok(()),
        }
    }

    /// Decode an inbound video/call frame and surface it to subscribers.
    /// Layout (build_video_frame): [16] call_id | [4] seq LE | [8] ts LE | [1] keyframe | [N] data.
    async fn handle_video_frame(&self, packet: &MeshPacket) -> Result<(), String> {
        let p = &packet.payload;
        if p.len() < 29 {
            return Err("video frame too short".to_string());
        }
        let call_id = Uuid::from_slice(&p[0..16]).map_err(|e| e.to_string())?;
        let seq = u32::from_le_bytes(p[16..20].try_into().unwrap());
        let timestamp_ms = i64::from_le_bytes(p[20..28].try_into().unwrap());
        let is_keyframe = p[28] != 0;
        let data = p[29..].to_vec();
        let _ = self.frame_received_tx.send(VideoFrameReceivedEvent {
            call_id,
            sender_uhid: packet.source_uhid.clone(),
            seq,
            timestamp_ms,
            is_keyframe,
            data,
        });
        Ok(())
    }

    // ── private ────────────────────────────────────────────────────────────────

    async fn send_signaling(&self, to_uhid: &str, sig: &VideoSignalingMessage) {
        let body = match serde_json::to_vec(sig) {
            Ok(b) => b,
            Err(_) => return,
        };
        let route = self.routing.find_route(to_uhid).await;
        let next_hop = route
            .map(|r| r.next_hop_uhid)
            .unwrap_or_else(|| to_uhid.to_string());

        let mut pkt = MeshPacket::new(PacketType::VideoSignaling, self.sender.local_uhid());
        pkt.destination_uhid = to_uhid.to_string();
        pkt.ttl = DEFAULT_TTL;
        pkt.priority = 32;
        pkt.payload = body;
        self.sender.send(&pkt, &next_hop).await;
    }

    async fn handle_signaling(&self, packet: &MeshPacket) -> Result<(), String> {
        let sig: VideoSignalingMessage = serde_json::from_slice(&packet.payload)
            .map_err(|e| format!("bad video signaling payload: {}", e))?;

        let mut calls = self.calls.lock().unwrap();
        match sig.kind.as_str() {
            "offer" => {
                calls.insert(
                    sig.call_id,
                    VideoCallEntry {
                        call_id: sig.call_id,
                        remote_uhid: packet.source_uhid.clone(),
                        state: VideoCallState::Incoming,
                    },
                );
            }
            "answer" => {
                if let Some(entry) = calls.get_mut(&sig.call_id) {
                    if entry.state == VideoCallState::Outgoing {
                        entry.state = VideoCallState::Connected;
                    }
                }
            }
            "hangup" => {
                if let Some(entry) = calls.get_mut(&sig.call_id) {
                    entry.state = VideoCallState::Ended;
                }
            }
            // keyframe_request and quality_change are surfaced to the app layer
            _ => {}
        }
        Ok(())
    }
}

// ── Helpers ───────────────────────────────────────────────────────────────────

static VIDEO_FRAME_SEQ: std::sync::atomic::AtomicU32 = std::sync::atomic::AtomicU32::new(0);

fn build_video_frame(call_id: Uuid, encoded_video: &[u8], is_keyframe: bool) -> Vec<u8> {
    let seq = VIDEO_FRAME_SEQ.fetch_add(1, std::sync::atomic::Ordering::Relaxed);
    let ts = unix_millis();
    let mut buf = Vec::with_capacity(29 + encoded_video.len());
    buf.extend_from_slice(call_id.as_bytes()); // [16]
    buf.extend_from_slice(&seq.to_le_bytes()); // [4]
    buf.extend_from_slice(&ts.to_le_bytes()); // [8]
    buf.push(if is_keyframe { 1 } else { 0 }); // [1]
    buf.extend_from_slice(encoded_video); // [N]
    buf
}

fn unix_millis() -> i64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_millis() as i64
}
